x-pack/plugins/security/server/authorization/privileges/feature_privilege_iterator/feature_privilege_iterator.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { Feature, FeatureKibanaPrivileges } from '../../../../../features/server';
+import { subFeaturePrivilegeIterator } from './sub_feature_privilege_iterator';
+
+interface IteratorOptions {
+  augmentWithSubFeaturePrivileges: boolean;
+  predicate?: (privilegeId: string, privilege: FeatureKibanaPrivileges) => boolean;
+}
+
+export function* featurePrivilegeIterator(
+  feature: Feature,
+  options: IteratorOptions
+): IterableIterator<{ privilegeId: string; privilege: FeatureKibanaPrivileges }> {
+  if (feature.privileges === 'none') return;
+
+  for (const entry of Object.entries(feature.privileges)) {
+    const [privilegeId, privilege] = entry;
+
+    if (options.predicate && !options.predicate(privilegeId, privilege)) {
+      continue;
+    }
+
+    if (options.augmentWithSubFeaturePrivileges) {
+      yield { privilegeId, privilege: mergeWithSubFeatures(privilegeId, privilege, feature) };
+    } else {
+      yield { privilegeId, privilege };
+    }
+  }
+}
+
+function mergeWithSubFeatures(
```

Comment:
Whenever we change `FeatureKibanaPrivileges` to have a new property, we'll have to remember to modify this function to properly merge... What do you think about turning `FeatureKibanaPrivileges` into a class and having a `merge` method?

Reply:
I'm not opposed to this change. If we go this route, I'd want to introduce a `FeatureKibanaPrivilegesConfig` or `RawFeatureKibanaPrivileges` interface so that feature authors can still specify "just" an object during feature registration, without having to construct a class instance.

Reply:
That makes sense to me. As long as this change doesn't make things super complicated, I think it'd help us long-term.

Reply:
That seems reasonable.